release-23.2: sql: maintain index dependencies during truncate

spilchen opened the pull request:
Backport 1/1 commits from #146287 on behalf of @spilchen.

----

When a table is truncated, new IDs are assigned to all its indexes. However, we previously failed to update any existing back-references to these index IDs in the table descriptor. This prevented the truncate from succeeding.

This change ensures that all references to old index IDs are properly remapped to the new ones, which allows the truncate to succeed.

Fixes #146065

Epic: none
Release note (bug fix): Fixed a bug that prevented TRUNCATE from succeeding if any indexes on the table had back-reference dependencies, such as from a view or function referencing the index.

----

Release justification: low-risk bug fix that unblocks TRUNCATE

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/146322)
<!-- Reviewable:end -->

spilchen committed:
sql: maintain index dependencies during truncate

When a table is truncated, new IDs are assigned to all its indexes.
However, we previously failed to update any existing back-references to these
index IDs in the table descriptor. This prevented the truncate from
succeeding.

This change ensures that all references to old index IDs are properly
remapped to the new ones, which allows the truncate to succeed.

Fixes #146065

Epic: none
Release note (bug fix): Fixed a bug that prevented TRUNCATE from
succeeding if any indexes on the table had back-reference dependencies,
such as from a view or function referencing the index.
